### PR TITLE
Skip some before_actions when processing payment intent from Stripe

### PR DESCRIPTION
**Pull request by andrewpbrett**
#### What? Why?

Closes https://github.com/openfoodfoundation/openfoodnetwork/issues/7692

When returning from Stripe, we were previously checking for a `current_distributor` and `current_order_cycle`; we're able to process the payment intent without these, and we still check for other things, like the OrderCycle being open and there being sufficient stock. 

#### What should we test?

I believe you could trigger this by changing your order cycle in a different window while SCA-authorizing the payment in another window; that's most likely not how the bug was happening in reality though. 

#### Release notes

Fixed a bug where a payment could be taken in Stripe but not marked as complete in OFN

Changelog Category: User facing changes

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
> I believe you could trigger this by changing your order cycle in a different window while SCA-authorizing the payment in another window; that's most likely not how the bug was happening in reality though.

Hmmm... agreed that's probably not how this is happening. But... when someone authorizes a payment for a **subscription**, they end up in this same code path right? That would almost certainly result in this issue appearing, as it's based on a mismatch between the state of their current session and the payment being processed, right?

It would be really nice if we could decouple this Stripe-redirect-processing from the main checkout controller actions somehow. Maybe not in this PR though. Just thinking out loud; could we change the Stripe redirect URL and have it processed by a separate controller that's designed specifically for handling Stripe responses? What do you think?

File: app/controllers/checkout_controller.rb [hunk 1]
```diff
@@ -150,10 +161,13 @@ def before_payment
...
-    @order.state == "payment" &&
-      last_payment&.state == "requires_authorization" &&
-      last_payment&.response_code == params["payment_intent"]
+    return false unless @payment = Spree::Payment.where(
+      response_code: params["payment_intent"],
+      state: "requires_authorization"
+    ).first
+
+    @order ||= @payment.order
+    @order.state == "payment" && @payment.order == @order
```
Comment (Matt-Yorkley) [inline app/controllers/checkout_controller.rb:170]: The second half of this line will fail / return false if the payment's order isn't the order that's currently loaded in the user's browser session, right? And then it won't be processed? :thinking:

**Comment by Matt-Yorkley** [comment]
So just to rubber-duck this a bit, the two theoretical scenarios where this could happen are:

1. The user is checking out and gets sent to Stripe to authorize the payment, and then for some reason in another tab (or something) they navigate to a different shop (which will recreate a new order/cart object), and then they authorize the payment in the other tab, and when they get redirected back to the checkout, the order in their session doesn't match the one that's being processed.
2. The user has a subscription that gets triggered at the end of the month. They need to authorize the payment and get an email, because it's off-session. After they authorize it they are redirected to the checkout, but the order object in their browser session doesn't match (different shop, different OC, whatever) and it fails to be processed correctly.

Is that roughly what we're looking at?

File: app/controllers/checkout_controller.rb (hunk 1 above)
Comment (andrewpbrett) [reply app/controllers/checkout_controller.rb:170]: I think that snuck in there by mistake, I was trying out and squashing a bunch of different things. I think it's better to process the payment intent regardless of the current order actually, I'll push a change.

**Comment by andrewpbrett** [comment]
> But... when someone authorizes a payment for a subscription, they end up in this same code path right?

No, there's a `return_url` param that we send to Stripe. In the case of authorizing a payment for a subscription, we direct them to the OrdersController and handle the payment intent there. 

It might make sense to unify these two, or it might not. I had the same thought you did here:

> could we change the Stripe redirect URL and have it processed by a separate controller

and started down that road a bit, but there are a lot of reasons to actually keep it in the CheckoutController: many of the before_actions are the same, for example, and we are, essentially, just doing one final update to the `Checkout` to say that the payment either succeeded or didn't. 

To answer your other questions (from "So just to rubber-duck this a bit..."):

Scenario 1 is the best guess that I can come up with for how this is happening, but it seems very unlikely. 
Scenario 2 isn't related to this, it's handled separately (see above).

This PR is strictly responding to what I saw in the logs, which was that there were multiple instances where someone was returning from Stripe, having authorized the payment there, trying to load the CheckoutController, and they were then redirected before OFN could confirm the payment intent, which was leading to the scenario described in the issue. For whatever reason, it was these two methods (`require_distributor_chosen` and `require_order_cycle`) that were doing the redirect; in my mind, we shouldn't have to require these things in order to confirm the payment intent.

**Review by Matt-Yorkley — APPROVED** [review]

**Comment by Matt-Yorkley** [comment]
Cool :ok_hand: :

> multiple instances where someone was returning from Stripe, having authorized the payment there, trying to load the CheckoutController, and they were then redirected before OFN could confirm the payment intent

That's super weird! I guess this could be related to the `www` thing? Redirecting back to the other domain?

**Comment by lin-d-hop** [comment]
@andrewpbrett No that does not sound like a common scenario

**Comment by sigmundpetersen** [comment]
Should we move this forward with one review due to its severity?
@jibees do yoy have any input/review on this?

**Review by jibees — APPROVED** [review]
Not familiar with it, but LGTM! 👌

**Comment by filipefurtad0** [comment]
Hey @andrewpbrett ,

After staging this PR I've gone through:

1) the same steps as [described in the issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/7692#issuecomment-902857548), and got the same result: the payment gets authorized and the order remains in the `payment` state - not in the complete state:

image

image

I'm not sure this mirrors the controller tests added to this PR. What do you think?

While this is probably not the root cause of this bug, it is interesting to note that the authorized but not captured payments from these testes (on the 20th of August) were automatically cancelled:
image


2) what seems to be the most likely cause of this scenario: shopping in different tabs

By opening the same session in the same browser, on separate windows it was possible to recreate this situation (with this PR staged):

image

However, the only way I was able to reach this state was by clicking the authorize payment for different carts almost simultaneously -> but I don't think this matches a real situation...

I don't think it is possible to recreate the `www` issue in a staging environment. Staging servers only accept the address `staging.` while in production we have the possibility to mismatch this.

It is perhaps a good idea to revisit this second test case, when #8127 is test ready.

**Summary**
With the tests above I was not able to probe any difference in behavior. Is this expected? Do you think there is anything else which should be tested here? Added the feedback label.

**Comment by andrewpbrett** [comment]
Thanks for testing this @filipefurtad0. Based on what I wrote [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/7692#issuecomment-912743458), I think it's better to close this, since the non-www / www redirect is more likely to be the actual cause.

**Pull request closed without merging**